### ISSUE TITLE
Some packages failed to build w/ LTO and `-Wl,--as-needed`

### ISSUE THREAD

**Issue by pchome** [InBetweenNames/gentooLTO#50]
I noticed `media-libs/mesa-17.2.3` failed to build while ` media-libs/mesa-17.2.2` was ok. Almost nothing changed in configs since last build and the only thing I remembered was `-Wl,--as-needed` accidentally (?) missed when preparing system for LTO.

So a while ago `LDFLAGS="-Wl,--hash-style=gnu ${CFLAGS}"` was copy-pasted from https://github.com/InBetweenNames/gentooLTO/blob/06cdb45ebfeddb6a91ec1d9fbf8dc611549db280/sys-config/ltoize/files/make.conf and then whole system was rebuilt w/ this. Until I noticed `LDFLAGS="-Wl,--as-needed -Wl,--hash-style=gnu"` in current `make.conf.lto` so copy-pasted into my make.conf again.

Busted:
```sh
libtool: link: x86_64-pc-linux-gnu-g++ -m32 -fvisibility=hidden -Werror=pointer-arith -Werror=vla -march=native -O3 -pipe -fgraphite-identity -ftree-loop-distribution -floop-nest-optimize -flto=3 -fuse-linker-plugin -Wall -fno-math-errno -fno-trapping-math -Wl,--as-needed -Wl,--hash-style=gnu -march=native -O3 -pipe -fgraphite-identity -ftree-loop-distribution -floop-nest-optimize -flto=3 -fuse-linker-plugin -o glsl_compiler glsl/main.o  glsl/.libs/libstandalone.a -lz -lpthread -pthread
/usr/lib/gcc/x86_64-pc-linux-gnu/7.2.0/32/libstdc++.so: undefined reference to `pthread_create'
/usr/lib/gcc/x86_64-pc-linux-gnu/7.2.0/32/libstdc++.so: undefined reference to `pthread_once'
/usr/lib/gcc/x86_64-pc-linux-gnu/7.2.0/32/libstdc++.so: undefined reference to `pthread_join'
collect2: error: ld returned 1 exit status
make[4]: *** [Makefile:2111: glsl_compiler] Error 1
make[4]: Leaving directory '/var/tmp/portage/media-libs/mesa-17.2.3/work/mesa-17.2.3-abi_x86_32.x86/src/compiler'
```

I'm played a bit w/ this command directly in this directory and removing `-Wl,--as-needed` from LDFLAGS did the trick.

I'm not sure is this related to LTO or to this https://bugs.gentoo.org/show_bug.cgi?id=as-needed bug list or misplaced library list in command.

**Comment by InBetweenNames** [comment]
Hah--I wish I checked this issue sooner!  I was actually just experimenting with this.  I noticed that removing LTO does indeed make mesa build again.  I suspect there's something going on inside the Mesa build system where the "as needed" bug only manifests under LTO conditions.  I bet it's some kind of corner case that makes the `--as-needed` "throw out" pthread early since it deems it isn't needed, and a later object requires that dependency, resulting in a link failure.  I rebuilt over 500 packages today and mesa is the only one that I've found so far exhibiting this error, so I'll commit a workaround specifically for "as needed" here.
